### PR TITLE
Small Android binary that extracts and loads the Perfetto producer.

### PR DESCRIPTION
**Pull request by pmuetschard**
Adds a small Android native binary, packaged as part of gapid.apk, which will look for the driver APK and load the GPU Perfetto data producer from it.

### CODE REVIEW

File: core/os/android/profiling/cc/launch_producer.cc [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      std::abort();                                 \
+    }                                               \
+    trans;                                          \
+  } while (0)
+#define GET_STRING_PROP(n, t) GET_PROP(n, t = _v)
+
+typedef void (*FN_PTR)(void);
+
+const char* kDriverProperty = "ro.gfx.driver.1";
+const char* kProducerSO = "/assets/libgpudataproducer.so";
```
Comment (lpy) [inline core/os/android/profiling/cc/launch_producer.cc]: We can't hardcode this, eventually this .so file is in lib/{ABI}/libgpudataproducer.so. Can a binary tell the ABI of itself?

Comment (pmuetschard) [reply core/os/android/profiling/cc/launch_producer.cc]: Done.

File: core/os/android/profiling/cc/launch_producer.cc [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+  GET_STRING_PROP(kDriverProperty, driver);
+  LOG_INFO("Driver package: %s", driver.c_str());
+  if (driver.empty()) {
+    LOG_ERR("No driver package set");
+    std::abort();
+  }
+
+  for (char const& c : driver) {
+    if (c == ' ' || c == '"' || c == '\'') {
```
Comment (AWoloszyn) [inline core/os/android/profiling/cc/launch_producer.cc]: Could we maybe get a comment saying "WHY" these characters are a problem?

File: core/os/android/profiling/cc/launch_producer.cc [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  std::string so = path + "!" + lib;
+  LOG_INFO("Trying %s", so.c_str());
+  void *handle = dlopen(so.c_str(), RTLD_GLOBAL);
+  if ((error = dlerror()) != nullptr || handle == nullptr) {
+    LOG_WARN("Error loading lib: %s", error);
+    return nullptr;
+  }
+
+  FN_PTR startFunc = (FN_PTR)dlsym(handle, "start");
```
Comment (AWoloszyn) [inline core/os/android/profiling/cc/launch_producer.cc]: Is the starting function actually called "start". That seems ... uninformative.

File: core/os/android/profiling/cc/launch_producer.cc [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  std::string path = getApkPath(driver);
+
+  dlerror();
+  FN_PTR startFunc = nullptr;
+  for (int i = 0; startFunc == nullptr && i < NELEM(kProducerPaths); i++) {
+    startFunc = loadLibrary(path, kProducerPaths[i]);
+  }
+
+  if (startFunc == nullptr) {
+    LOG_ERR("Did not found the producer library");
```
Comment (AWoloszyn) [inline core/os/android/profiling/cc/launch_producer.cc]: Did not find the producer library.

File: core/os/android/profiling/cc/launch_producer.cc (hunk 3 above)
Comment (pmuetschard) [reply core/os/android/profiling/cc/launch_producer.cc]: My comment exactly. Waiting on a decision in the design doc. If it gets changed, will update the code accordingly.

File: core/os/android/profiling/cc/launch_producer.cc (hunk 2 above)
Comment (pmuetschard) [reply core/os/android/profiling/cc/launch_producer.cc]: Done. I've also improved the check to be more strict.

**Comment by lpy** [comment]
Quick update: So far we will stick with `start` as the entry point name.

**Pull request merged**